lib/blockscore/connection.rb

```diff
@@ -7,60 +7,51 @@ module Connection
     USER_AGENT    = "blockscore-ruby/#{BlockScore::VERSION} (#{REPO})".freeze
     CONTENT_TYPE  = 'application/json'.freeze
     API_URL       = URI.parse('https://api.blockscore.com').freeze
+    HEADERS       = {
+      'Accept'       => ACCEPT_HEADER,
+      'User-Agent'   => USER_AGENT,
+      'Content-Type' => CONTENT_TYPE
+    }.freeze
 
-    def get(path, params)
-      request(:get, path, params)
+    def delete(path, _)
+      request(:delete, path, nil)
     end
 
-    def post(path, params)
-      request(:post, path, params)
-    end
+    private
 
-    def patch(path, params)
-      request(:patch, path, params)
+    def get(path, params)
+      request(:get, encode_path_params(path, params), nil)
     end
 
-    def delete(path, params)
-      request(:delete, path, params)
+    def post(path, params)
+      request(:post, path, params.to_json)
     end
 
-    private
-
-    def headers
-      @headers ||= {
-        'Accept'       => ACCEPT_HEADER,
-        'User-Agent'   => USER_AGENT,
-        'Content-Type' => CONTENT_TYPE
-      }
+    def patch(path, params)
+      request(:patch, path, params.to_json)
     end
 
-    def request(method, endpoint, params)
-      fail NoAPIKeyError, 'No API key was provided.' unless BlockScore.api_key
-
-      path = API_URL + endpoint.to_s
-
+    def request(method, final_endpoint, params)
       begin
-        response = execute_request(method, path, params)
+        response = execute_request(method,
+                                   API_URL + final_endpoint.to_s,
+                                   params)
       rescue SocketError, Errno::ECONNREFUSED => e
-        fail APIConnectionError, e.message
+        fail APIConnectionError, e
```

Comment:
> Unfortunately the only thing retained is the message.

You can retain the backtrace by passing it in as the 3rd argument to raise, eg:

```ruby
fail APIConnectionError, e, e.backtrace
```

Also, I would recommend against single letter variable names, even in this specific instance.